### PR TITLE
Working update for 1.17

### PR DESCRIPTION
**Pull request by MisterSoandSo**
Closes issues #228 and #229 for getting a basic working 1.17 client.

### CODE REVIEW

**Comment by MisterSoandSo** [comment]
I'll be working on trying to implement the Microsoft login api while referencing #218 for 1.17+/2021~ support

File: minecraft/networking/packets/clientbound/play/map_packet.py [hunk 1]
```diff
@@ -100,7 +101,10 @@ def read(self, file_object):
             icon = MapPacket.MapIcon(type, direction, (x, z), display_name)
             self.icons.append(icon)
 
-        self.width = UnsignedByte.read(file_object)
+        try:
+            self.width = UnsignedByte.read(file_object)
+        except:
+            self.width = None
```
Comment (joodicator) [inline minecraft/networking/packets/clientbound/play/map_packet.py:107]: What is the reason for this? If the `width` read was raising an exception, I feel like there's probably a better way of dealing with that.

**Comment by joodicator** [comment]
Thank you, @MisterSoandSo and other contributors, for your work on 1.17 and later versions. I'm prepared to merge these changes as a basis for some additional changes I'll add afterwards, but, first, please see my one specific question about the code.

With regards to your comment elsewhere:
> I plan to try rewrite Pycraft to support 1.17+ as I feel that back porting to previous versions would make things harder down the line.

That is probably a good idea, as the current codebase is burdened by a need to support older versions of the API, and by some suboptimal API decisions that were made along the way, and development has slowed down a lot due to a lack of maintainers able and willing to keep it properly up to date. Personally, I think I will retire from the project in full after seeing that pyCraft supports 1.18.1 for my own purposes. We can discuss with the repository owner what steps can be taken then with regards to your rewrite of pyCraft.

File: minecraft/networking/packets/clientbound/play/map_packet.py (hunk 1 above)
Comment (MisterSoandSo) [reply minecraft/networking/packets/clientbound/play/map_packet.py:107]: Ya that was me be young and lazy to find out the issue when I played on a earlier mc version. I intend to fix that as that is a situational patch that plagued me on a server that I played on.

**Comment by MisterSoandSo** [comment]
> 

At the moment, I'm try a few things with my current project [ Protocol Database](https://github.com/MisterSoandSo/Minecraft-Packet-Protocol-Database) to help improve maintainability  with each version. Once I figure out how I would like to reference protocols with JSON. Then I would backwards with the current pycraft library and depreciate content pre1.16.4 and the old Mojang auth.  Layman terms alot of rewiring before making it neat.

File: minecraft/networking/packets/clientbound/play/map_packet.py (hunk 1 above)
Comment (joodicator) [reply minecraft/networking/packets/clientbound/play/map_packet.py:107]: Maybe the packet format is being parsed incorrectly for earlier versions. Do you remember which version it was?

**Comment by joodicator** [comment]
> At the moment, I'm try a few things with my current project [ Protocol Database](https://github.com/MisterSoandSo/Minecraft-Packet-Protocol-Database) to help improve maintainability with each version. Once I figure out how I would like to reference protocols with JSON.

This is definitely something that was sorely missing from pyCraft. Updating protocol information manually is not very scalable. In case you're not aware, there are theree existing projects that attempt to do something vaguely similar to this:
* <https://github.com/PrismarineJS/minecraft-data>, which seems promising, but I haven't looked at it in detail;
* <https://github.com/joodicator/mc-dev-data> which is what am currently using to update pyCraft, and works by parsing [wiki.vg]() with manual intervention to patch errors, but is rather slow and hard to maintain, so could stand to be replaced with something better;
* <https://github.com/Pokechu22/Burger>, which extracts protocol data from Minecraft's decompiled Java code, but IIRC the protocol part isn't working at the moment, so fixing that is an available task.

Perhaps you can take inspiration from these, assuming you make your own version - it could be that a combination of the approaches used above is best.

File: minecraft/networking/packets/clientbound/play/map_packet.py (hunk 1 above)
Comment (MisterSoandSo) [reply minecraft/networking/packets/clientbound/play/map_packet.py:107]: Let me get back to you on that as this was mainly a fix and forget type of situation as it seems like a random inconsistent bug. I can play for hours back then and the bug wouldn't show up. I suspect the server I played on might be sending malformed packets or something to that nature.

**Pull request merged**